code/controllers/subsystem/move_loop.dm

```diff
@@ -0,0 +1,466 @@
+
+/**
+ * Wrapper around walk_rand(), doesn't actually result in a random walk, it's more like moving to random places in viewish
+ *
+ * Arguments:
+ * moving - The atom we want to move
+ * delay - How many seconds to wait between fires. Defaults to the lowest value, 0.1
+ * timeout - Time in seconds until the moveloop self expires. Defaults to infinity
+ * override - Should we replace the current loop if it exists. Defaults to TRUE
+ *
+ * Returns TRUE if the loop sucessfully started, or FALSE if it failed
+**/
+/datum/controller/subsystem/movement_loop/proc/move_to_rand(moving, delay, timeout, override)
+	return start_looping(/datum/move_loop/move_to_rand, override, moving, delay, timeout)
+
+///Template class of the walk() replacements, handles the timing portion of the loops
+/datum/move_loop
+	///The thing we're moving about
+	var/atom/movable/moving
+	///Lifetime in seconds, defaults to forever
+	var/lifetime = INFINITY
+	///Delay between each move in seconds
+	var/delay = 0.1
+	///We use this to track the delay between movements
+	var/timer = 0
+	///The last tick we processed
+	var/lasttick = 0
+
+/datum/move_loop/proc/setup(atom/moving, delay = 0.1, timeout = INFINITY)
+	if(!ismovable(moving))
+		handle_delete()
+		return FALSE
+
+	src.moving = moving
+	src.delay = delay
+	lifetime = timeout
+
+	RegisterSignal(moving, COMSIG_PARENT_QDELETING, .proc/handle_delete)
+	return TRUE
+
+/datum/move_loop/proc/kill()
+	SHOULD_CALL_PARENT(TRUE)
+	if(moving)
+		SEND_SIGNAL(moving, COMSIG_MOVELOOP_END)
+		UnregisterSignal(moving, COMSIG_PARENT_QDELETING)
+
+/datum/move_loop/proc/handle_delete()
+	SHOULD_CALL_PARENT(TRUE)
+	SSmovement_loop.remove_from_loop(moving, src)
+
+/datum/move_loop/process(delta_time)
+	timer = round(timer + delta_time, 0.1) //Round up due to floating point shit
+	if(timer >= lifetime)
+		handle_delete()
+		return
+	if(round(timer - delay, 0.1) < lasttick)
+		return
+	if(SEND_SIGNAL(moving, COMSIG_MOVELOOP_PROCESS_CHECK) & MOVELOOP_STOP_PROCESSING) //Chance for the object to react
+		return
+
+	lasttick = timer
+	move()
+
+///Handles the actual move, overriden by children
+/datum/move_loop/proc/move()
+	return
+
+///Replacement for walk()
+/datum/move_loop/move
+	var/direction
+
+/datum/move_loop/move/setup(atom/moving, delay, timeout, dir)
+	if(!..())
+		return FALSE
+	direction = dir
+	return TRUE
+
+/datum/move_loop/move/move()
+	moving.Move(get_step(moving, direction), direction)
+
+
+/datum/move_loop/has_target
+	///The thing we're moving in relation to, either at or away from
+	var/atom/target
+
+/datum/move_loop/has_target/setup(atom/moving, delay, timeout, atom/chasing)
+	if(!..())
+		return FALSE
+	if(!isatom(chasing))
+		handle_delete()
+		return FALSE
+
+	target = chasing
+
+	if(!isturf(target))
+		RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/handle_no_target) //Don't do this for turfs, because of reasons
+
+	return TRUE
```

Comment:
```suggestion
/datum/move_loop/has_target/setup(atom/moving, delay, timeout, atom/chasing)
	. = ..()
	if(!.)
		return
	if(!isatom(chasing))
		handle_delete()
		return FALSE

	target = chasing

	if(!isturf(target))
		RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/handle_no_target) //Don't do this for turfs, because of reasons

```